quickstart/backend/src/main/java/com/digitalasset/quickstart/ledger/LedgerApi.java

```diff
@@ -194,19 +238,26 @@ CompletableFuture<Result> exerciseAndGetResult(
                         .build();
         currentSpan.addEvent("built ledger submit request");
 
-        // Submit and decode the exercise result
         return toCompletableFuture(commands.submitAndWaitForTransactionTree(request))
                 .thenApply(response -> {
                     currentSpan.addEvent("received ledger submit response");
 
-                    // The "rootEventIds" should have exactly one root event for a single command
                     String eventId = response.getTransaction().getRootEventIds(0);
                     currentSpan.setAttribute("backend.submit.response.event.id", eventId);
 
                     TransactionOuterClass.TreeEvent event = response.getTransaction().getEventsByIdMap().get(eventId);
                     ValueOuterClass.Value resultPayload = event.getExercised().getExerciseResult();
-                    // Convert the result payload back to the type <Result>
+
                     return (Result) proto2Dto.choiceResult(choice.templateId(), choice.choiceName()).convert(resultPayload);
+                })
+                .whenComplete((res, ex) -> {
+                    if (ex != null) {
+                        logger.error("Failed to exercise choice='{}' on contractId='{}' as party='{}': {}",
+                                choice.choiceName(), contractId.getContractId, party, ex.getMessage(), ex);
+                    } else {
+                        logger.info("Successfully exercised choice='{}' on contractId='{}' as party='{}' with commandId='{}'",
```

Comment:
We have a full result here, so we should definitely be logging transactionId, probably offset (less important). Adding at least the workflow and submission ids would also be worth doing.